crypto/asn1/a_time.c

```diff
@@ -519,31 +517,26 @@ int ossl_asn1_time_print_ex(BIO *bp, const ASN1_TIME *tm, unsigned long flags)
         }
 
         if ((flags & ASN1_DTFLGS_TYPE_MASK) == ASN1_DTFLGS_ISO8601) {
-            return BIO_printf(bp, "%4d-%02d-%02d %02d:%02d:%02d%.*s%s",
-                          stm.tm_year + 1900, stm.tm_mon + 1,
-                          stm.tm_mday, stm.tm_hour,
-                          stm.tm_min, stm.tm_sec, f_len, f,
-                          (gmt ? "Z" : "")) > 0;
-        }
-        else {
-            return BIO_printf(bp, "%s %2d %02d:%02d:%02d%.*s %d%s",
-                          _asn1_mon[stm.tm_mon], stm.tm_mday, stm.tm_hour,
-                          stm.tm_min, stm.tm_sec, f_len, f, stm.tm_year + 1900,
-                          (gmt ? " GMT" : "")) > 0;
+            return BIO_printf(bp, "%4d-%02d-%02d %02d:%02d:%02d%.*sZ",
```

Comment:
Although you're not changing the fractional part of the of the format string. It seems wrong:
> `%.*s`

This won't put a period into the output, as would be expected with fractional seconds, but is used to indicate the supplied length is a _precision_ value (rather than the minimum field length).

Reply:
Indeed, this code only replaces the last %s with a fixed "Z", not changing anything in/meddling with the format logic otherwise.... So what's your suggestion how to change?

Reply:
It is not wrong for ASCII machines because the ASCII `.` is included in `f`.

However for EBCDIC and ASCII `.` will be printed instead of EBCDIC `.`.

Reply:
I'm not sure what you're asking for @t8m: Do you want me to change the pre-existing logic to handle fractional seconds differently? I don't quite understand where this PR introduced an error (?).

Reply:
>It is not wrong for ASCII machines because the ASCII `.` is included in `f`.

I see that now. EBCDIC would require changing the input (i.e. `tm`) or making a copy.

Reply:
> I don't quite understand where this PR introduced an error

It did not introduce an error, it was an preexisting error of this code on EBCDIC systems when printing the fractional seconds. But as you are already touching these formats, you could also fix it.

Reply:
Ah, OK. Then please allow some time for me to learn (about EBCDIC?) to even understand the bug... Any hints about a test environment on Linux welcome.

Reply:
I'd be willing to let the generation of `f` be fixed separately.

Reply:
> Ah, OK. Then please allow some time for me to learn (about EBCDIC?) to even understand the bug... Any hints about a test environment on Linux welcome.

There is no EBCDIC linux. You would have to have s390x machine and run openssl directly in Z/OS. You do not want to do that :grin: 

Basically `.` in the print out of the fractional seconds would have to be part of the format string and not taken from the ASN1_TIME string as it currently is.

Reply:
Hmpf. Then this is basically "blind" coding? Taking the "." out of "f" indeed seems simple. Proposal in latest commit. If NOK, just exclude https://github.com/openssl/openssl/pull/26344/commits/cb619bbe97c0971bb776a536607c32714922bf85 from the PR.

Reply:
On second thought: Are you sure changing the "." to be printed via `printf` is enough? If the fractional seconds string (incl. ".") is ASCII, so should be the digits, no? Stated another way: Doesn't this demand that also the fractional second digits need to be "re-coded"? Strange old world... Can't we simply suggest Z/OS users to get themselves something slightly more up-to-date? :-) And/or leave this as-is: If those guys didn't notice the wrong encoding for the last few years, they won't notice a fix either.